catalog/core/catalog-core-attributeregistry/src/main/java/ddf/catalog/data/impl/AttributeRegistryImpl.java

```diff
@@ -56,14 +61,50 @@ public boolean register(final AttributeDescriptor attributeDescriptor) {
     @Override
     public void deregister(final String name) {
         notNull(name, "The attribute name cannot be null.");
-
         attributeMap.remove(name);
     }
 
     @Override
     public Optional<AttributeDescriptor> lookup(final String name) {
         notNull(name, "The attribute name cannot be null.");
-
         return Optional.ofNullable(attributeMap.get(name));
     }
+
+    public void setBundleContext(BundleContext bundleContext) {
+        this.bundleContext = bundleContext;
+    }
+
+    public BundleContext getBundleContext() {
+        return bundleContext;
+    }
+
+    public void bind(ServiceReference ref) {
+        if (bundleContext != null) {
+            MetacardType metacardType = (MetacardType) bundleContext.getService(ref);
+            registerMetacardTypeAttributes(metacardType);
+        }
+    }
+
+    public void unbind(ServiceReference ref) {
+        if (bundleContext != null) {
+            MetacardType metacardType = (MetacardType) bundleContext.getService(ref);
+            deregisterMetacardTypeAttributes(metacardType);
+        }
+    }
+
+    private void registerMetacardTypeAttributes(MetacardType metacardType) {
+        LOGGER.debug("Adding attributes from {} metacard type", metacardType.getName());
+        metacardType.getAttributeDescriptors()
+                .stream()
+                .forEach(this::register);
+
+    }
+
+    private void deregisterMetacardTypeAttributes(MetacardType metacardType) {
+        LOGGER.debug("Removing attributes from {} metacard type", metacardType.getName());
+        metacardType.getAttributeDescriptors()
+                .stream()
+                .map(attributeDescriptor -> attributeDescriptor.getName())
+                .forEach(this::deregister);
```

Comment:
This could lead to attributes being removed from the registry when they shouldn't be.  If 2 metacardTypes attempt to register the same attribute, only 1 will be registered. If 1 metacardType is removed - it will remove the registry entry for both types.